sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update.hpp

```diff
@@ -0,0 +1,364 @@
+    return false;
+  }
+  int ii = dense_ind(ind, VL, cfg);
+  bool res = (ii % VL) != cfg.masked_lane;
+  return res;
+}
+
+// ----------------- Actual "traits" for each operation.
+
+template <class T, int N, class C, C Op> struct ImplIncBase {
+  static constexpr C atomic_op = Op;
+  static constexpr int n_args = 0;
+
+  static T init(int i, const Config &cfg) { return (T)0; }
+
+  static T gold(int i, const Config &cfg) {
+    T gold = is_updated(i, N, cfg)
+                 ? (T)(cfg.repeat * cfg.threads_per_group * cfg.n_groups)
+                 : init(i, cfg);
+    return gold;
+  }
+};
+
+template <class T, int N, class C, C Op> struct ImplDecBase {
+  static constexpr C atomic_op = Op;
+  static constexpr int n_args = 0;
+  static constexpr int base = 5;
+
+  static T init(int i, const Config &cfg) {
+    return (T)(cfg.repeat * cfg.threads_per_group * cfg.n_groups + base);
+  }
+
+  static T gold(int i, const Config &cfg) {
+    T gold = is_updated(i, N, cfg) ? (T)base : init(i, cfg);
+    return gold;
+  }
+};
+
+// The purpose of this is validate that floating point data is correctly
+// processed.
+constexpr float FPDELTA = 0.5f;
+
+template <class T, int N, class C, C Op> struct ImplLoadBase {
+  static constexpr C atomic_op = Op;
+  static constexpr int n_args = 0;
+
+  static T init(int i, const Config &cfg) { return (T)(i + FPDELTA); }
+
+  static T gold(int i, const Config &cfg) {
+    T gold = init(i, cfg);
+    return gold;
+  }
+};
+
+template <class T, int N>
+struct ImplLoad : ImplLoadBase<T, N, atomic_op, atomic_op::load> {};
+template <class T, int N>
+struct ImplInc : ImplIncBase<T, N, atomic_op, atomic_op::inc> {};
+template <class T, int N>
+struct ImplDec : ImplDecBase<T, N, atomic_op, atomic_op::dec> {};
+
+// ----------------- Main function and test combinations.
+
+template <int N, template <class, int> class Op, bool UseProperties,
+          int SignMask = (Signed | Unsigned)>
+bool test_int_types(queue q, const Config &cfg) {
+  bool passed = true;
+  if constexpr (SignMask & Signed) {
+    // Supported by LSC atomic:
+    if constexpr (UseProperties)
+      passed &= test<int16_t, N, Op, UseProperties>(q, cfg);
+
+    passed &= test<int32_t, N, Op, UseProperties>(q, cfg);
+  }
+
+  if constexpr (SignMask & Unsigned) {
+    // Supported by LSC atomic:
+    if constexpr (UseProperties)
+      passed &= test<uint16_t, N, Op, UseProperties>(q, cfg);
+
+    passed &= test<uint32_t, N, Op, UseProperties>(q, cfg);
+  }
+  return passed;
+}
+
+template <int N, template <class, int> class Op, bool UseProperties>
+bool test_fp_types(queue q, const Config &cfg) {
+  bool passed = true;
+  passed &= test<float, N, Op, UseProperties>(q, cfg);
+  if (q.get_device().has(sycl::aspect::atomic64) &&
+      q.get_device().has(sycl::aspect::fp64)) {
+    // Disable double data type for fcmpxchg operation as D64 data is not
+    // supported for that operation.
```

Comment:
This comment is unclear. Does it say that fadd needs to be tested, but fcmpxchg - not? If so, I couldn't find the code skipping fcmpxchg for D64. Is it TODO comment?

Reply:
More like TODO, I removed it for now. Will add this when I'll get to two-operand atomic.